src/libraries/Microsoft.Extensions.Logging/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -0,0 +1,659 @@
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Extensions.Logging.Generators
+{
+    public partial class LoggerMessageGenerator
+    {
+        internal class Parser
+        {
+            private readonly CancellationToken _cancellationToken;
+            private readonly Compilation _compilation;
+            private readonly Action<Diagnostic> _reportDiagnostic;
+
+            public Parser(Compilation compilation, Action<Diagnostic> reportDiagnostic, CancellationToken cancellationToken)
+            {
+                _compilation = compilation;
+                _cancellationToken = cancellationToken;
+                _reportDiagnostic = reportDiagnostic;
+            }
+
+            /// <summary>
+            /// Gets the set of logging classes containing methods to output.
+            /// </summary>
+            public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSyntax> classes)
+            {
+                const string LoggerMessageAttribute = "Microsoft.Extensions.Logging.LoggerMessageAttribute";
+
+                INamedTypeSymbol loggerMessageAttribute = _compilation.GetTypeByMetadataName(LoggerMessageAttribute);
+                if (loggerMessageAttribute == null)
+                {
+                    // nothing to do if this type isn't available
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol loggerSymbol = _compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.ILogger");
+                if (loggerSymbol == null)
+                {
+                    // nothing to do if this type isn't available
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol logLevelSymbol = _compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.LogLevel");
+                if (logLevelSymbol == null)
+                {
+                    // nothing to do if this type isn't available
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol exceptionSymbol = _compilation.GetTypeByMetadataName("System.Exception");
+                if (exceptionSymbol == null)
+                {
+                    Diag(DiagnosticDescriptors.MissingRequiredType, null, "System.Exception");
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol enumerableSymbol = _compilation.GetTypeByMetadataName("System.Collections.IEnumerable");
+                if (enumerableSymbol == null)
+                {
+                    Diag(DiagnosticDescriptors.MissingRequiredType, null, "System.Collections.IEnumerable");
+                    return Array.Empty<LoggerClass>();
+                }
+
+                INamedTypeSymbol stringSymbol = _compilation.GetTypeByMetadataName("System.String");
```

Comment:
Prefer `_compilation.GetSpecialType(SpecialType.System_String)` here. That will return the `System.String` from mscorlib.